web/src/composebox_typeahead.ts

```diff
@@ -642,27 +736,84 @@ export function get_sorted_filtered_items(query, input_element) {
 
     const big_results = get_candidates(query, input_element);
 
-    if (!big_results) {
-        return false;
+    if (big_results === false) {
+        return [];
     }
 
-    const opts = get_stream_topic_data(input_element);
-
-    if (completing === "mention" || completing === "silent_mention") {
-        return filter_and_sort_mentions(big_results.is_silent, token, opts);
+    // These are sorted separately
+    if (big_results.type === "mention") {
+        return big_results.list;
     }
 
-    return filter_and_sort_candidates(completing, big_results, token);
+    assert(big_results.type === completing);
+    return filter_and_sort_candidates(big_results, token);
 }
 
-export function filter_and_sort_candidates(completing, candidates, token) {
-    const matcher = compose_content_matcher(completing, token);
+type PillItem =
+    | UserGroupPillData
+    | UserOrMentionPillData
+    | TopicJumpPillData
+    | TimeJumpPillData
+    | LanguagePillData
+    | TopicPillData
+    | EmojiPillData
+    | SlashCommandPillData
+    | StreamPillData;
+
```

Comment:
So both the `PillIData` types and `List` types have a `type` field, and this is because we often work with the lists and I want to be able to easily get what kind of list it is. But when an item is selected in the typeahead we just have the individual item and we need to know the type from there too. This feels a little janky but I couldn't think of something better.